src/protocol/sort/multi_bit_permutation.rs

```diff
@@ -0,0 +1,256 @@
+            }
+        }
+        equality_check_futures.push(async move {
+            // multiply all mult_input for this j for each record => f(j)
+            multiply_all_shares(
+                ctx_across_bits,
+                RecordId::from(idx * num_records + rec),
+                mult_input.as_slice(),
+            )
+            .await
+        });
+    }
+    try_join_all(equality_check_futures).await
+}
+
+/// Get binary representation of an integer as a vector of bool
+fn get_binary_from_int(input_num: usize, num_bits: usize) -> Vec<bool> {
+    let mut num = input_num;
+    let mut bits = Vec::with_capacity(num_bits);
+    while num != 0 {
+        bits.push(num & 1 == 1);
+        num >>= 1;
+    }
+    bits.resize_with(num_bits, Default::default);
+    bits.reverse();
+    bits
+}
+
+#[cfg(all(test, not(feature = "shuttle")))]
+mod tests {
+    use futures::future::try_join_all;
+
+    use crate::{
+        ff::{Field, Fp31},
+        protocol::{
+            context::Context,
+            sort::{
+                multi_bit_permutation::{
+                    get_binary_from_int, get_bit_equality_checkers, multi_bit_permutation,
+                },
+                MultiBitPermutationStep::EqualityBitChecker,
+            },
+        },
+        test_fixture::{Reconstruct, Runner, TestWorld},
+    };
+    const INPUT: [&[u128]; 3] = [
+        &[1, 0, 1, 0, 1, 0],
+        &[0, 1, 1, 0, 0, 0],
+        &[0, 0, 1, 0, 1, 0],
+    ];
+    const EXPECTED: &[u128] = &[4, 3, 6, 1, 5, 2]; //100 010 111 000 101 000
+    const EXPECTED_NUMS: &[usize] = &[4, 2, 7, 0, 5, 0];
+    #[tokio::test]
+    pub async fn semi_honest() {
+        let world = TestWorld::new().await;
+
+        let input: Vec<Vec<_>> = INPUT
+            .into_iter()
+            .map(|v| v.iter().map(|x| Fp31::from(*x)).collect())
+            .collect();
+        let result = world
+            .semi_honest(input, |ctx, m_shares| async move {
+                multi_bit_permutation(ctx, m_shares.as_slice())
+                    .await
+                    .unwrap()
+            })
+            .await;
+
+        assert_eq!(&result.reconstruct(), EXPECTED);
+    }
+
+    #[tokio::test]
+    pub async fn bit_equality_checkers() {
+        let world = TestWorld::new().await;
+
+        let input: Vec<Vec<_>> = INPUT
+            .into_iter()
+            .map(|v| v.iter().map(|x| Fp31::from(*x)).collect())
+            .collect();
+
+        let num_multi_bits = INPUT.len();
+        let num_records = INPUT[0].len();
+        let num_possible_bit_values = 2 << (num_multi_bits - 1);
+
+        let result = world
+            .semi_honest(input, |ctx, m_shares| async move {
+                let mut equality_check_futures =
+                    Vec::with_capacity(num_possible_bit_values * num_records);
+                for j in 0..num_possible_bit_values {
+                    let ctx_equality_checker = ctx.narrow(&EqualityBitChecker);
```

Comment:
I don't think this narrow call is necessary, because within `get_bit_equality_checkers` you're immediately calling narrow again:
`let ctx_across_bits = ctx.narrow(&MultiplyAcrossBits);`
So as far as I can tell, you never actually use this `ctx_equality_checker` context on its own.